crates/pathfinder/src/sequencer/builder.rs

```diff
@@ -0,0 +1,388 @@
+            BlockId::Hash(hash) => ("blockHash", hash.0.to_hex_str()),
+            // These have to use "blockNumber", "blockHash" does not accept tags.
+            BlockId::Latest => ("blockNumber", Cow::from("latest")),
+            BlockId::Pending => ("blockNumber", Cow::from("pending")),
+        };
+
+        self.add_param(name, &value)
+    }
+
+    pub fn with_contract_address(self, address: ContractAddress) -> Self {
+        self.add_param("contractAddress", &address.0.to_hex_str())
+    }
+
+    pub fn with_class_hash(self, class_hash: ClassHash) -> Self {
+        self.add_param("classHash", &class_hash.0.to_hex_str())
+    }
+
+    pub fn with_optional_token(self, token: Option<&str>) -> Self {
+        match token {
+            Some(token) => self.add_param("token", token),
+            None => self,
+        }
+    }
+
+    pub fn with_storage_address(self, address: StorageAddress) -> Self {
+        use crate::rpc::serde::starkhash_to_dec_str;
+        self.add_param("key", &starkhash_to_dec_str(&address.0))
+    }
+
+    pub fn with_transaction_hash(self, hash: StarknetTransactionHash) -> Self {
+        self.add_param("transactionHash", &hash.0.to_hex_str())
+    }
+
+    pub fn add_param(mut self, name: &str, value: &str) -> Self {
+        self.url.query_pairs_mut().append_pair(name, value);
+        self
+    }
+
+    pub fn auto_retry(self) -> Request<'a, WithRetry> {
+        Request {
+            url: self.url,
+            client: self.client,
+            marker: PhantomData::default(),
+        }
+    }
+
+    pub fn without_retry(self) -> Request<'a, WithoutRetry> {
+        Request {
+            url: self.url,
+            client: self.client,
+            marker: PhantomData::default(),
+        }
+    }
+}
+
+impl<'a> Request<'a, WithoutRetry> {
+    pub async fn get<T>(self) -> Result<T, SequencerError>
+    where
+        T: serde::de::DeserializeOwned,
+    {
+        let response = self.client.get(self.url).send().await?;
+        parse::<T>(response).await
+    }
+
+    pub async fn get_as_bytes(self) -> Result<bytes::Bytes, SequencerError> {
+        let response = self.client.get(self.url).send().await?;
+        let bytes = parse_raw(response).await?.bytes().await?;
+        Ok(bytes)
+    }
+
+    pub async fn post_with_json<T, J>(self, json: &J) -> Result<T, SequencerError>
+    where
+        T: serde::de::DeserializeOwned,
+        J: serde::Serialize + ?Sized,
+    {
+        let response = self.client.post(self.url).json(json).send().await?;
+        parse::<T>(response).await
+    }
+}
+
+impl<'a> Request<'a, WithRetry> {
+    pub async fn get<T>(self) -> Result<T, SequencerError>
+    where
+        T: serde::de::DeserializeOwned,
+    {
+        let without_retry = Request::<WithoutRetry> {
+            url: self.url,
+            client: self.client,
+            marker: PhantomData::default(),
+        };
+
+        retry0(|| async move { without_retry.get().await }, retry_condition).await
```

Comment:
I was hoping to get this to work, but am unable to as retry (and its foundation `TokioRetryIf`) require a `FutureFactory`. This is not possible because my entire framework consists of `Url` which is not `Copy`.